### PR TITLE
Do not apply smart quotes if Sphinx says no, even if Docutils says yes

### LINKED ISSUE #4359: A global ``.docutils`` smart_quotes setting overrides html_use_smartypants or extension setting
Consider situation [mentioned](https://github.com/sphinx-doc/sphinx/issues/3967#issuecomment-339408120) by @matthew-brett: user has a ``~/.docutils`` file which activates or deactivates smart_quotes.

Then this will override Sphinx's own setting either from (deprecated)  html_use_smartypants, or from an extension which would initiate a `smart_quotes` setting (cf ee52c7cd4f5e0a300c8fdd897c9529f5c633c7ed). This is bad because it is to be expected that a local setting overrides a global one.

Also, the setting in a `docutils.conf` file will override one set by an extension despite ee52c7cd.

Thus if we want to deactivate by default smart quotes for some builders (man, text, cf #4357), via a simple-minded usage of a `'smart_quotes'` setting as in ``environment/__init__.py`` this will fail is this issue is not fixed first.

One can test this by creating a ``$HOME/.doctutils`` file with

```
[parsers]
	smart_quotes=no
```

and use `html_use_smartypants = True` in `conf.py`. No smart quotes will be applied.

Or set ``smart_quotes=yes`` in Docutils user config file and `html_use_smartypants = False` in `conf.py` of local Sphinx project then smart quotes will be applied.

### Environment info
- OS: Mac
- Python version: 3.6.4
- Sphinx version: 1.6.5, stable, master

Relates #4112

### PR DESCRIPTION
**Pull request by jfbu**
This closes #4359.

Relates #3967.

### CODE REVIEW

File: sphinx/io.py
Comment (jfbu) [inline sphinx/io.py]: as far as I understand, this will be called from a BuildEnvironment object ``read_doc()`` method, when a ``reader`` is being set up, and at that time the ``app.env.settings['smart_quotes']`` always exists with value either True or False due to earlier code executed by the BuildEnvironment instance.

Now, perhaps the accessor method is superfluous here and we should set the Boolean directly rather than having such an instance method added to RSTParser class.

**Comment by jfbu** [comment]
@tk0miya perhaps we can reconsider now the 79df05b36 which reverted bfd39c12b ? I forgot the reason why bfd39c12b was reverted. With this PR, a Sphinx `smart_quotes` conf.py setting will be obeyed and override whatever Docutils global config file says. It will also override local ``docutils.conf`` so we should modify our docs.

edit1: "override" is not correct; it applies only in case the `smart_quotes` is activated via ``docutils.conf``, then if Sphinx decides to turn it off, it will work. But it `smart_quotes` is de-activated (explicitely) via ``docutils.conf``, then no matter what Sphinx config says, it will remain de-activated. In brief, de-activation whether coming from Docutils or Sphinx side always wins.

edit2: I have ressurected bfd39c12b in modified form in 2nd commit of this PR.

**Comment by jfbu** [comment]
I should clarify that initial commit of this PR does only what the title says: if Sphinx setting disallows smart quotes, then they are not applied and Docutils setting is ignored. But if Sphinx setting does allow smart quotes, then they still can be unactivated if Docutils configuration files ``docutils.conf`` or ``~/.docutils`` says so.

File: sphinx/parsers.py
Comment (tk0miya) [inline sphinx/parsers.py]: Let's make `RSTParser` class a subclass of `sphinx.parsers.Parser`. After that we can refer config object by `self.config`.
I think storing status to class variable is not good way. Because it behaves like global variables.

File: doc/config.rst
Comment (tk0miya) [inline doc/config.rst]: I agree the basic concept of this PR. But we would need to be able to disable smart_quotes by several condition.
For example, #4142 tells us to disable it for Japanese document. I saw somebody wants to disable it for arbitrary builder in some issues.
So I think this should not be boolean. (but I don't have good alternative idea...)

**Comment by tk0miya** [comment]
>perhaps we can reconsider now the 79df05b which reverted bfd39c1 ?

The reason of the revert is only it is not intended commit. It was not my completed work as I commented above. I think simple boolean setting is not good for this case. So we have to consider the way to configure smart_quotes.

It's okay to introduce a configuration for smart_quotes itself to me.

**Comment by tk0miya** [comment]
I'd forggeten to mention about parsers. I think smart quotes is not only for reST files. It is a setting for representation, so it should effect to other formats.
If user want to do that, it is better if configurable.

File: sphinx/parsers.py
Comment (jfbu) [reply sphinx/parsers.py]: agreed, done (will comment after push)

File: sphinx/parsers.py
Comment (jfbu) [inline sphinx/parsers.py]: I have a problem here which is that afaik SphinxSmartQuotes is a class, not an instance of that class, and in get_transforms below it is the class object which is added to list of transforms. Hence I needed to modify the class attribute here. Don't know if this can have undesired consequences.

File: sphinx/environment/__init__.py
Comment (jfbu) [inline sphinx/environment/__init__.py]: Docutils has no smartquotes_action user config setting yet; but it introduced the docutils.transforms.universal.SmartQuotes class attribute at 0.14, and it is useful for us, because Sphinx users may want to for example not apply smart quotes to ellipsis conversion (from  ``...``). For Docutils < 0.14, there was at some point some bad config which educated backticks also, and we override that in our own sphinx.util.smartypants (this module is used by Sphinx only for Docutils < 0.14).

**Comment by jfbu** [comment]
> I'd forggeten to mention about parsers. I think smart quotes is not only for reST files. It is a setting for representation, so it should effect to other formats.
If user want to do that, it is better if configurable.

My knowledge of how Sphinx interacts with Docutils remains quite limited. At 69cf328b you created a class ``RSTParser`` apparently especially for facilitating customization at Sphinx level of Docutils Smart Quotes transform. Do you mean here that this code should actually be in Sphinx Parser class, not RSTParser?

**Comment by jfbu** [comment]
>  I think simple boolean setting is not good for this case. So we have to consider the way to configure smart_quotes.

@tk0miya I have added confval setting ``smartquotes_excludes``. It has limitations particularly for builders, as multiple builders can share same pickled environment, so user has to explicitely request re-parsing of source files when doing html then text builds for example; or when doing `make html text` then smart quotes will be applied to text target too, despite setting of ``smartquotes_excludes``. Best I can do for the moment due to shared BuildEnvironment.

I have not addressed the issue about extending PR to other than RSTParser, because my knowledge of Sphinx/Docutils workflow is currently too limited.

File: doc/config.rst
Comment (jfbu) [reply doc/config.rst]: I have addressed that by supplementary `smartquotes_excludes` dict configuration setting at 88a6e2451 (fixed typo in doc at 4b3a09133)

File: doc/config.rst
Comment (jfbu) [inline doc/config.rst]: ah sorry, I am not used to ``sphinx-build`` but as it creates by default a ``.doctrees`` individual to each target, there is no need for ``-E``. The problem arises when using the "make-mode" only. Will fix docs.

File: sphinx/environment/__init__.py
Comment (jfbu) [inline sphinx/environment/__init__.py]: now that Sphinx core has a ``smart_quotes`` config setting, possibly the logic for

    if 'smart_quotes' not in self.settings:

(see #4112) is gone? can an extension override a config setting rather than add it to environment as done at  https://github.com/sphinx-contrib/spelling/commit/6209fa8d2c16182b30053f40770dc81b11c77f00 ?

**Comment by jfbu** [comment]
Currently the sphinx-quickstart created Makefile (whether make-mode enabled or not) configures builds to share `doctrees/` across build targets. On the other hand the default behaviour of `sphinx-build` does not do that. I see the speed gain into sharing the cached files, but the smart quotes issue shows that there is a problem when the Docutils parsing should behave differently depending on builder. It is a bit of thorny issue because indeed it will be disconcerting to user if ``make html epub latex man`` re-parses source files (which takes time on big docs) for each target. We can not escape that by implementing "man will always force re-parsing", because then ``make man html`` will inhibit smart quotes in html output. However we could set "man" and "text" builders to force re-parsing *and* store the parsed files in special location. I.e. there would be ``doctrees/`` and ``doctrees-nosmartquotes/``.

Then the builders listed in ``smartquotes_excludes['builder']`` would use ``doctrees-nosmartquotes`` and the others ``doctrees/``. This looks a bit ugly for only one problem and not very scalable in (improbable) case some other Docutils implemented transform than Smart Quotes arises in future with similar issues.

File: doc/config.rst
Comment (tk0miya) [inline doc/config.rst]: I think it is better to rename keys plural form.

File: sphinx/environment/__init__.py
Comment (tk0miya) [reply sphinx/environment/__init__.py]: I think it is not a public interface, but it is needed for workaround. So please keep this as is.

Note: It might be refactored in future. Personally, I feel `read_doc` is not a part of responsibility of "BuildEnvironment".

File: sphinx/environment/__init__.py
Comment (tk0miya) [inline sphinx/environment/__init__.py]: Ideally, smart quotes transform should be invoked after the pickling (write phase) to support each builders.
I know it is a bit big refactoring. so +1 this for stable.

File: sphinx/parsers.py
Comment (tk0miya) [reply sphinx/parsers.py]: Right, class attribute is a kind of global variable. So modifying it is not good way. So it would be better to update it temporarily like following:
```
from contextlib import contextmanager

@contextmanager
def sphinx_smartquotes_action(env):
    try:
        original = SphinxSmartQuotes.smartquotes_action
        SphinxSmartQuotes.smart_quotes_action = env.config.smartquotes_action
        yield
     finally:
         SphinxSmartQuotes.smart_quotes_action = original

# in env.doc_read() method:
with sphinx_smartquotes_action(self):
    doctree = read_doc(...)
```

File: sphinx/environment/__init__.py
Comment (tk0miya) [reply sphinx/environment/__init__.py]: Is there any reason for setting this into `self.settings`?

**Comment by tk0miya** [comment]
>Do you mean here that this code should actually be in Sphinx Parser class, not RSTParser?

docutils picks up transform modules by `get_components()` of related components. So I think `SphinxStandaloneReader` is better.

File: sphinx/environment/__init__.py
Comment (jfbu) [reply sphinx/environment/__init__.py]: It was in anticipation that Docutils might offer such an option in future to configure the `smartquotes_action` attribute of its docutils.transforms.universal.SmartQuotes, which we inherit for SphinxSmartQuotes. Also, I did this before RSTParser inherited from Parser hence got access to app. Thinking twice, it might indeed be better not to set this into ``self.settings``, else we will again have problem that a Docutils config file would override the Sphinx user conf.py, if such an option is introduced in future by Docutils.

File: sphinx/parsers.py
Comment (jfbu) [reply sphinx/parsers.py]: Thanks, will do.

File: sphinx/environment/__init__.py
Comment (jfbu) [reply sphinx/environment/__init__.py]: ok, I will leave it. I find "BuildEnvironment" complicated...

File: doc/config.rst
Comment (jfbu) [reply doc/config.rst]: I was using singular for easing up re-use of the key as ``getattr(self.config, key)``, but I agree it is more self-explanatory to use plurals. In theory any config key is usable here to condition usage of Smart Quotes, but in practice I guess I can just drop the final "s". Or we allow only conditioning on language and builder? Also, I will explain better in docs that the logic is simply "OR".

File: sphinx/environment/__init__.py
Comment (tk0miya) [reply sphinx/environment/__init__.py]: If docutils does not implement the option yet, it might be too premature action. -1 for adding key.

File: sphinx/environment/__init__.py
Comment (tk0miya) [reply sphinx/environment/__init__.py]: Absolutely, `BuildEnvironment` has multiple responsibility. So I'd like to split it into some modules in future.

**Comment by jfbu** [comment]
@tk0miya I have implemented your advices. I checked that SmartQuotes then applies also to Markdown files as added following [the docs](http://www.sphinx-doc.org/en/stable/markdown.html).

> If docutils does not implement the option yet, it might be too premature action. -1 for adding key.

I removed it ;-) (I mean, I do not add it to `env.settings` but I keep it as a Sphinx option, as it is useful)

File: sphinx/environment/__init__.py
Comment (jfbu) [reply sphinx/environment/__init__.py]: I have removed it from ``self.settings``, but I do keep it as confval because it is useful for Sphinx users. Is that ok?

File: doc/config.rst
Comment (tk0miya) [reply doc/config.rst]: I can't think of the usecase for any config key (without `language`). So I feel it's okay to restrict the key to `languages` and `builders`.

>Also, I will explain better in docs that the logic is simply "OR".

+1

File: sphinx/parsers.py
Comment (jfbu) [inline sphinx/parsers.py]: now inheriting from Parser is not needed, because we don't need access to `app`. Should I delete that added inheritance ?

File: sphinx/io.py
Comment (jfbu) [inline sphinx/io.py]: with this, also ``.md`` source files (for example) are in the scope of Smart Quotes, which was not the case before.

File: sphinx/environment/__init__.py
Comment (jfbu) [inline sphinx/environment/__init__.py]: perhaps the whole ``else`` clause could be dropped if we limit to conditioning on language and builder

File: doc/config.rst
Comment (jfbu) [inline doc/config.rst:409]: should we document like this our defects ;-) or leave it to user to discover ;-) ?

File: sphinx/environment/__init__.py
Comment (tk0miya) [inline sphinx/environment/__init__.py]: Is there any reason to using `setattr`? An assignment is much simpler to me.

File: sphinx/io.py
Comment (tk0miya) [inline sphinx/io.py]: Please override `get_transforms()` and add `SphinxSmartQuotes` conditionally. because `self.transforms` is a class variable. So `SphinxSmartQuotes` will be added on each instantiation.
```
def __init__(...):
    ...
    self.smart_quotes = app.env.settings['smart_quotes']

def get_transforms(self):
    transforms = SphinxBaseReder.get_transforms()
    if self.smart_quotes:
        transforms.append(SphinxSmartQuotes)

    return transforms
```

File: sphinx/parsers.py
Comment (tk0miya) [inline sphinx/parsers.py]: `Parser` super class is no longer needed.

**Review by tk0miya — CHANGES_REQUESTED** [review]
LGTM with nits

File: sphinx/io.py
Comment (jfbu) [reply sphinx/io.py]: sorry for my basic mistake. Indeed build time considerably increased due to it. I will fix.

File: doc/config.rst
Comment (tk0miya) [reply doc/config.rst:409]: This is internally and hard to understand behavior. So it should be documented. Let's go with as is.

File: sphinx/io.py
Comment (tk0miya) [reply sphinx/io.py]: :-)

File: sphinx/environment/__init__.py
Comment (tk0miya) [reply sphinx/environment/__init__.py]: As I commented above, I can't imagine the use case for this else block.
I don't opposite it, but I'd like to know an example use case before merging.

File: sphinx/environment/__init__.py
Comment (jfbu) [reply sphinx/environment/__init__.py]: I used an assignment but it did not have expected effect in my testing. Was it because SphinxSmartQuotes is imported separately in io.py?

Comment (jfbu) [reply sphinx/environment/__init__.py]: hmm... works perfectly well with an assignment. Don't know what went wrong (but I am doing various things at the same time, bad habit...). Ok, will push fix.

File: sphinx/parsers.py
Comment (jfbu) [reply sphinx/parsers.py]: done

File: sphinx/environment/__init__.py
Comment (jfbu) [inline sphinx/environment/__init__.py]: the try is needed because Docutils SmartQuotes does not have the ``smartquotes_action`` attribute at Docutils < 0.14.

File: sphinx/parsers.py
Comment (jfbu) [inline sphinx/parsers.py]: The Parser inheritance was added in an earlier commit of this PR (#4360), but is now not needed

File: sphinx/environment/__init__.py
Comment (jfbu) [reply sphinx/environment/__init__.py]: sorry I overlooked your comment, I will remove that block as I don't see obvious use case either

File: sphinx/environment/__init__.py
Comment (tk0miya) [inline sphinx/environment/__init__.py]: This is not executed when the error has happened at yield block. Usually, it does not effect to the processing because `sphinx-build` crashes on the error. But the behavior might be changed in future. So it should be write back correctly using `final` clause.
How about this implementation?
```
if not hasattr(SphinxSmartQuotes, 'smartquotes_action'):
    # less than docutils-0.14
    yield
else:
    # docutils-0.14 or above
    try:
        original = SphinxSmartQuotes.smartquotes_action
        SphinxSmartQuotes.smartquotes_action = env.config.smartquotes_action
    finally:
        SphinxSmartQuotes.smartquotes_action = original
```

File: sphinx/environment/__init__.py
Comment (tk0miya) [inline sphinx/environment/__init__.py:96]: Please add type annotation: `# type: (BuildEnvironment) -> Generator`.

**Review by tk0miya — CHANGES_REQUESTED** [review]

File: sphinx/environment/__init__.py
Comment (tk0miya) [reply sphinx/environment/__init__.py]: +1

File: sphinx/environment/__init__.py
Comment (jfbu) [reply sphinx/environment/__init__.py]: +1, I guess I add a ``yield`` in the ``try`` ;-)

**Comment by jfbu** [comment]
closing and reopening to trigger again testing

**Review by tk0miya — APPROVED** [review]
Great work!

File: CHANGES
Comment (tk0miya) [inline CHANGES]: I noticed `smartquotes_disables` does not mentioned here.

Anyway, these settings for smart quotes are named `smart_quotes` or `smartquotes`. I feel they should have same prefix. What do you think?

Comment (jfbu) [reply CHANGES]: `smartquotes_excludes` is mentioned at next line ...

I agree about using same prefix `smartquotes` for all. Docutils has a `smart_quotes` or `smart-quotes` option and this why I originally used `smart_quotes`. But Docutils has the `smartquotes_action` class attribute hence I also followed. But I feel `smartquotes` prefix for all conf.py variables is much better indeed.

Comment (tk0miya) [reply CHANGES]: oh, sorry > `smartquotes_excludes`

**Review by tk0miya — APPROVED** [review]
LGTM!

**Comment by tk0miya** [comment]
ping @jfbu 

Do you still have any remaining work for this PR? If none, could you merge this please?
Then I will release 1.6.6 in this weekend.

If you have tasks, I can wait it. So please let me know.

Thanks,

**Comment by jfbu** [comment]
@tk0miya Thanks for very thorough reviewing and expert Python guidance! To make for cleaner commit history and due to the fact the PR was importantly refactored mid-way I will rebase and squash commits and then I will merge. ;-)

(I am preserving for some time original history [there](https://github.com/jfbu/sphinx/tree/4360_before_squash))

**Comment by jfbu** [comment]
I will merge once testing completes :tada:

**Comment by jfbu** [comment]
Merging because Travis fails due to connect timeout errors.

**Comment by tk0miya** [comment]
🎉  Great work!

**Pull request merged**
